File: mg01/game/tech/Assets/Script/Game/Test/KUI/Pop/TestPop4.cs
/* ==============================================================================
 * TestPop4
 * @author jr.zeng
 * 2018/9/8 11:37:01
 * ==============================================================================*/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;


using UnityEngine;
using Object = UnityEngine.Object;

using mg.org;
using mg.org.KUI;


public class TestPop4 : KUIPop
{

    KButton m_btnOk;
    KButton m_btnClose;
    //
    KInputField m_inputName;
    KText m_labelTestInput;
    //
    KToggle m_tgl_1;
    KText m_labelTgl_1;
    KToggleGroup m_tglGroup_1;
    //
    KSlider m_sliderSchedule;
    KText m_labelSchedule;
    //
    KProgressBar m_barLoading;
    //
    KImage m_icon;
    //
    KScrollView m_scrollView;
    KListViewScroll m_listView;

    public TestPop4()
    {
        m_popId = POP_ID.TEST_POP_4;
        m_layerIdx = POP_LAYER_IDX.LAYER_POP_1;

        ShowGameObject();

        //
        m_btnOk = GetChildByName<KButton>("Button_Ok", true);
        m_btnClose = GetChildByName<KButton>("Button_Close", true);
        //
        m_inputName = GetChildByName<KInputField>("Input_name", true);
        m_labelTestInput = GetChildByName<KText>("Label_testInput", true);
        //
        m_tgl_1 = GetChildByName<KToggle>("Toggle_position1", true);
        m_labelTgl_1 = GetChildByName<KText>(m_tgl_1.gameObject, "Label_Text", true);
        m_tglGroup_1 = GetChildByName<KToggleGroup>("ToggleGroup_a1", true);
        //
        m_sliderSchedule = GetChildByName<KSlider>("Slider_Schedule1", true);
        m_labelSchedule = GetChildByName<KText>("Label_testSilder", true);
        //
        m_barLoading = GetChildByName<KProgressBar>("ProgressBar_loading", true);
        //
        m_icon = GetChildByName<KImage>("Image_sharedAnchor", true);
        //
        m_scrollView = GetChildByName<KScrollView>("ScrollView_GuildList", true);
        m_listView = ComponentUtil.EnsureComponent<KListViewScroll>(m_scrollView.gameObject);
    }

    protected override void __Show(object showObj_, params object[] params_)
    {
        ShowButton();
        ShowText();
        ShowToggle();
        ShowSilder();
        ShowProBar();
        ShowImage();
        ShowScrollView();
        ShowListView();

    }

    protected override void SetupEvent()
    {
        CCApp.keyboard.Attach(KEY_EVENT.RELEASE, OnKeyPress, this);

    }

    protected override void ClearEvent()
    {
        CCApp.keyboard.Detach(KEY_EVENT.RELEASE, OnKeyPress);

    }
    protected override void __Destroy()
    {
        ClearButton();
        ClearText();
        ClearToggle();
        ClearSilder();
        ClearProBar();
        ClearImage();
        ClearScrollView();
        ClearListView();

        CCApp.soundMgr.StopBgm();
    }

    void OnKeyPress(object evt_)
    {
        KeyCode key = (KeyCode)evt_;
        switch (key)
        {
            case KeyCode.Keypad1:
                m_listView.JumpToIndex(11, KListViewScroll.JumpPosType.BOTTOM);
                break;
            case KeyCode.Keypad2:
                m_listView.JumpToIndex(5, KListViewScroll.JumpPosType.CENTER);
                break;
            case KeyCode.Keypad3:
                m_listView.JumpToIndex(14, KListViewScroll.JumpPosType.TOP);
                break;
            default:
                break;
        }

    }

    
    //-------~★~-------~★~-------~★~测试按钮~★~-------~★~-------~★~-------//

    void ShowButton()
    {
        m_btnOk.onClick.AddListener(OnClickBtn);
        m_btnClose.onClick.AddListener(OnClickBtn);
    }


    void ClearButton()
    {
        m_btnOk.onClick.RemoveListener(OnClickBtn);
        m_btnClose.onClick.RemoveListener(OnClickBtn);
    }

    void OnClickBtn(KButton btn_)
    {
        if (btn_ == m_btnOk)
        {
            Log.Debug("Click OK", this);
        }
        else if(btn_ == m_btnClose)
        {
            Close();
        }

    }


    //-------~★~-------~★~-------~★~测试文本~★~-------~★~-------~★~-------//
    
    void ShowText()
    {
        m_inputName.onValueChanged.AddListener(OnInputValue);
        m_inputName.onEndEdit.AddListener(OnInputEndEdit);
    }


    void ClearText()
    {

        m_inputName.onValueChanged.RemoveListener(OnInputValue);
        m_inputName.onEndEdit.RemoveListener(OnInputEndEdit);
    }

    void OnInputValue(string str_)
    {
        m_labelTestInput.text = "编辑中: " + str_;
    }

    void OnInputEndEdit(string str_)
    {
        m_labelTestInput.text = "编辑完成: " + str_;
    }


    //-------~★~-------~★~-------~★~测试单/多选框~★~-------~★~-------~★~-------//

    void ShowToggle()
    {
        m_tgl_1.onValueChanged.AddListener(OnToggleValue);

        m_tgl_1.needReqChange = true;   //需要请求改变
        m_tgl_1.onReqChange.AddListener(OnToggleReq);

        m_tglGroup_1.allowMultiple = true;  //多选
        m_tglGroup_1.onValueChange.AddListener(OnTglGroupValue);
    }
    
    void ClearToggle()
    {
        m_tgl_1.onValueChanged.RemoveListener(OnToggleValue);
        m_tgl_1.onReqChange.RemoveListener(OnToggleReq);

        m_tglGroup_1.onValueChange.RemoveListener(OnTglGroupValue);
    }


    void OnToggleValue(KToggle tgl_, bool b_)
    {
        m_labelTgl_1.text = b_ ? "选中" : "没选中";
    }

    void OnToggleReq(KToggle tgl_, bool b_)
    {
        m_tgl_1.Select(b_);
    }


    void OnTglGroupValue(KToggleGroup tglGroup_, int idx_, bool b_)
    {
        KToggle child = tglGroup_.GetToggle(idx_);
        KText label = GetChildByName<KText>(child.gameObject, "Label_Text");

        label.text = b_ ? "选中" : "没选中";
    }


    //-------~★~-------~★~-------~★~测试滑块~★~-------~★~-------~★~-------//

    void ShowSilder()
    {
        m_sliderSchedule.onValueChanged.AddListener(OnSliderValue);
        m_sliderSchedule.value = 0.5f;

    }
    

    void ClearSilder()
    {
        m_sliderSchedule.onValueChanged.RemoveListener(OnSliderValue);

    }


    void OnSliderValue(KSlider tgl_, float value_)
    {
        m_labelSchedule.text = value_.ToString();
    }


    //-------~★~-------~★~-------~★~测试进度条~★~-------~★~-------~★~-------//

    void ShowProBar()
    {
        m_barLoading.value = 0.8f;
    }
    

    void ClearProBar()
    {

    }



    //-------~★~-------~★~-------~★~测试图片~★~-------~★~-------~★~-------//

    void ShowImage()
    {
        SprAtlasCache.me.LoadSprite(this, m_icon, "npc1032", "npc1032", true);
    }



    void ClearImage()
    {

    }


    //-------~★~-------~★~-------~★~测试ScrollView~★~-------~★~-------~★~-------//

    void ShowScrollView()
    {

    }



    void ClearScrollView()
    {

    }


    //-------~★~-------~★~-------~★~测试ListView~★~-------~★~-------~★~-------//

    void ShowListView()
    {
        LayoutParam layoutParam = m_listView.layoutParam;
        //layoutParam.dir = LayoutDirection.LeftToRight;
        //layoutParam.divNum = 2;
        layoutParam.origin = new Vector2(0, -200f);
        layoutParam.itemGap = new Vector2(0, 50f);
        //m_listView.direction = KScrollView.ScrollDir.horizontal;
        m_listView.onDataChanged.AddListener(UpdateListItem);

        //int len = MathUtil.RandomInt(10, 30);
        int len = 200;
        m_listView.ShowLen(len, 100);
    }



    void ClearListView()
    {
        m_listView.ClearList();
        m_listView.onDataChanged.RemoveListener(UpdateListItem);

    }

    void UpdateListItem(GameObject item_, int index_, object data_)
    {
        GameObject select = GetChildByName(item_, "Image_bg/select");
        GameObject normal = GetChildByName(item_, "Image_bg/normal1");
        KText labelName = GetChildByName<KText>(item_, "Label_Name");
        KText labelWinNum = GetChildByName<KText>(item_, "Label_WinNum");
        KText labelMemberNum = GetChildByName<KText>(item_, "Label_MemberNum");

        bool b = (index_ % 2) == 0;
        normal.SetActive(b);
        select.SetActive(!b);

        //index_ += 1;
        labelName.text = "路人甲" + index_;
        labelWinNum.text = index_.ToString();
        labelWinNum.text = 999 + "人";
    }
    

}

File: mg01/game/tech/Assets/Resources/LuaScript/src/game/test/pop/TestPop4.lua
-- TestPop4
--@author jr.zeng
--@date 2018/2/6  15:03
local modname = "TestPop4"
--==================global reference=======================

--===================namespace========================
local ns = "test"
local using = {"org.action"}
local _ENV = namespace(ns, using, modname)

--===================module property========================
local super = UIPop	--父类
TestPop4 = class(modname, super, _ENV)
local TestPop4 = TestPop4

--===================module content========================

local ui_root = "Container_TestPop4/Container_Panel"

local ui_map = {

	btn_ok = "Button_Ok",
	btn_close = "Button_Close",

	toggle_1 = {"Toggle_position1",
		label_toggle1 = "Label_Text"
	},

	tglgroup_1 = "ToggleGroup_a1",

	input_name = "Input_name",
	label_testInput = "Label_testInput",

	slider_schedule1 = "Slider_Schedule1",
	label_testSilder = "Label_testSilder",

	progbar_loading = "ProgressBar_loading",

	image_icon = "Container_Image/Image_sharedAnchor",

	sv_guildlist = {"ScrollView_GuildList",
		image_mask = "Image_mask",
		container_content = "Image_mask/Container_content",
	},
}


local ui_map_item =
{
	container_select = "Image_bg/select",
	container_normal1 = "Image_bg/normal1",
	label_name = "Label_Name",
	label_winNum = "Label_WinNum",
	label_memberNum = "Label_MemberNum",
}

TestPop4.pop_id = POP_ID.TEST_4

function TestPop4:__ctor()

	self.m_handler = self:get_handler(ui_root, ui_map)     --代理

	local go = self.m_handler.gameObject

	--self.m_handler.sv_guildlist:ensure_component(CTypeName.KListViewScroll)

end

function TestPop4:__show(showObj, ...)

	self:show_btn()
	self:show_text()
	self:show_toggle()
	self:show_slider()
	self:show_progbar()
	self:show_image()
	self:show_scrollview()
	self:show_listview()
	--
	self:show_res()
	--self:show_my_list()

end

function TestPop4:setup_event()

	App.keyboard:attach(KEY_EVENT.RELEASE, self.on_key_press, self)

end


function TestPop4:clear_event()

	App.keyboard:detach(KEY_EVENT.RELEASE, self.on_key_press, self)

end


function TestPop4:__destroy()

	self:clear_btn()
	self:clear_text()
	self:clear_toggle()
	self:clear_slider()
	self:clear_progbar()
	self:clear_image()
	self:clear_scrollview()
	self:clear_listview()
	--
	self:clear_res()
	--self:clear_my_list()
end


function TestPop4:on_key_press(evt_)

	--nslog.print_t("on_key_press", evt_.data)

	local code = evt_.data
	if code == KeyCode.Keypad1 then

		GameObjUtil.remove_from_parent(self.m_handler.btn_ok.gameObject)

	elseif code == KeyCode.Keypad2 then

		GameObjUtil.change_parent(self.m_handler.btn_ok.gameObject, self.m_handler.gameObject)
	end

end

--//-------~★~-------~★~-------~★~button~★~-------~★~-------~★~-------//


function TestPop4:show_btn()

	self.m_handler.btn_ok:attach_click(self.on_click_ok, self)
	self.m_handler.btn_close:attach_click(self.on_click_close, self)


end

function TestPop4:clear_btn()

	self.m_handler.btn_ok:detach_click(self.on_click_ok, self)
	self.m_handler.btn_close:detach_click(self.on_click_close, self)
end


local a = 1
function TestPop4:on_click_ok(go)

	--nslog.print_t('on_click_ok', self, go)
	self.m_handler.toggle_1.label_toggle1:set_text(a.."次")
	a = a + 1
end

function TestPop4:on_click_close(go)

	self:close()
end

--//-------~★~-------~★~-------~★~text~★~-------~★~-------~★~-------//

function TestPop4:show_text()

	self.m_handler.input_name:attach_input_change(self.on_input_change, self)
	self.m_handler.input_name:set_input_type( InputType.Password )

end


function TestPop4:clear_text()

	self.m_handler.input_name:detach_input_change(self.on_input_change, self)

end


function TestPop4:on_input_change(go, str)

	self.m_handler.label_testInput:set_text(str)
end


--//-------~★~-------~★~-------~★~toggle~★~-------~★~-------~★~-------//


function TestPop4:show_toggle()

	self.m_handler.toggle_1.label_toggle1:set_text("骄傲的勾选框")
	self.m_handler.tglgroup_1:allow_mult_tglgroup(true)


	self.m_handler.tglgroup_1:attach_tglgroup_change(self.on_toggle_group, self)
end


function TestPop4:clear_toggle()


	self.m_handler.tglgroup_1:detach_tglgroup_change(self.on_toggle_group, self)

end

function TestPop4:on_toggle_group(go, index, b)

	self.m_handler.toggle_1.label_toggle1:set_text((b and "选中 " or "取消 ") .. index)
end


--//-------~★~-------~★~-------~★~silder~★~-------~★~-------~★~-------//


function TestPop4:show_slider()

	self.m_handler.slider_schedule1:attach_slider_change(self.on_slider_change, self)

	self.m_handler.label_testSilder:set_text( self.m_handler.slider_schedule1:get_slider_value() )
end


function TestPop4:clear_slider()


	self.m_handler.slider_schedule1:detach_slider_change(self.on_slider_change, self)
end

function TestPop4:on_slider_change(go, value)

	self.m_handler.label_testSilder:set_text(value)
end

--//-------~★~-------~★~-------~★~progressbar~★~-------~★~-------~★~---

function TestPop4:show_progbar()

	local function getValue()
		return self.m_handler.progbar_loading:get_prog_value()
	end

	local function setValue(value)
		self.m_handler.progbar_loading:set_prog_value(value)
	end

	self.m_handler.progbar_loading:set_prog_value(0)

	local action1 = action.ease(
		action.set_to(4, {
			toValue = 1,
			getter = getValue,
			setter = setValue,
		}),
		EaseType.BounceIn
	)

	App.actionMgr:run(self, action1)
end


function TestPop4:clear_progbar()

	App.actionMgr:stop(self)
end

--//-------~★~-------~★~-------~★~image~★~-------~★~-------~★~-------//

function TestPop4:show_image()

	self.m_handler.image_icon:load_sprite('npc1031', true)

end


function TestPop4:clear_image()

end


--//-------~★~-------~★~-------~★~scrollview~★~-------~★~-------~★~-------//

function TestPop4:show_scrollview()

end


function TestPop4:clear_scrollview()

end


--//-------~★~-------~★~-------~★~klistview~★~-------~★~-------~★~-------//

function TestPop4:show_listview()

	local list_view = self.m_handler.sv_guildlist

	list_view:set_layout_param({
		itemGap = cc.p(20,30),
	})
	--self.m_handler.sv_guildlist:set_scrollable(false)

	local data
	local data_list = {}
	local len = 999
	for i=1, len do
		data = {idx=i, name="你是谁 "..i }
		data_list[#data_list+1] = data
	end

	list_view:show_list(data_list, self.update_item, 11)
end


function TestPop4:clear_listview()

	self.m_handler.sv_guildlist:clear_list()

end

function TestPop4:update_item(item, index, data)

	local handler = self:get_handler(item, ui_map_item)

	local b = index % 2 == 0
	handler.container_select:set_active(not b)
	handler.container_normal1:set_active(b)

	handler.label_name:set_text(data.name)
	handler.label_winNum:set_text(data.idx)
	handler.label_memberNum:set_text(99 .. "人")

end


--//-------~★~-------~★~-------~★~Resource~★~-------~★~-------~★~-------//


function TestPop4:show_res()

	local go_path = 'GUI/cn/Prefab/Canvas_Test3'
	--GameObjCache:load_async(go_path, self.on_go_loaded, self)

end

function TestPop4:clear_res()


end


function TestPop4:on_go_loaded(go, url)

	--nslog.print_t('on_go_loaded', go, url)

end

--//-------~★~-------~★~-------~★~klistview~★~-------~★~-------~★~-------//

function TestPop4:show_my_list()

	if not self.m_listView then

		self.m_listView = ListView.new({
			container = self.m_handler.sv_guildlist.container_content,
			divNum = 1,
			--on_item_data = callback(self.on_item_data, self),
			item_cls = TestListItem,

		})

		retain(self.m_listView, self)
	end


	local data
	local data_list = {}
	local len = 13
	for i=1, len do
		data = {idx=i, name="你是谁 "..i }
		data_list[#data_list+1] = data
	end

	self.m_listView:show(data_list)
end


function TestPop4:clear_my_list()

	if self.m_listView then
		self.m_listView:destroy()
	end

end

function TestPop4:on_item_data(item, index, data)

	local handler = self:get_handler(item, ui_map_item)

	local b = index % 2 == 0
	handler.container_select:set_active(not b)
	handler.container_normal1:set_active(b)

	handler.label_name:set_text(data.name)
	handler.label_winNum:set_text(data.idx)
	handler.label_memberNum:set_text(99 .. "人")

end


return TestPop4